tests/test_petsc.py

```diff
@@ -772,3 +773,252 @@ def test_time_loop():
 
     assert 'ctx0.t0 = t0' in body4
     assert body4.count('ctx0.t0 = t0') == 1
+
+
+class TestCoupledLinear:
+    # The coupled interface can be used even for uncoupled problems, meaning
+    # the equations will be solved within a single matrix system.
+    # These tests use simple problems to validate functionality, but they help
+    # ensure correctness in code generation.
+    # TODO: Add more comprehensive tests for fully coupled problems.
+    # TODO: Add subdomain tests, time loop, multiple coupled etc.
+
+    @skipif('petsc')
+    # NOTE: Coupled mode does not yet support multiple processors, but
+    # it still needs to be tested in a parallel environment
+    @pytest.mark.parallel(mode=[1])
+    def test_coupled_vs_non_coupled(self, mode):
+        grid = Grid(shape=(11, 11), dtype=np.float64)
+
+        e = Function(name='e', grid=grid, space_order=2, dtype=np.float64)
+        f = Function(name='f', grid=grid, space_order=2, dtype=np.float64)
+        g = Function(name='g', grid=grid, space_order=2, dtype=np.float64)
+        h = Function(name='h', grid=grid, space_order=2, dtype=np.float64)
+
+        f.data[:] = 5.
+        h.data[:] = 5.
+
+        eq1 = Eq(e.laplace, f)
+        eq2 = Eq(g.laplace, h)
+
+        # Non-coupled
+        petsc1 = PETScSolve(eq1, target=e)
+        petsc2 = PETScSolve(eq2, target=g)
+        with switchconfig(openmp=False, mpi=True):
+            op1 = Operator(petsc1 + petsc2, opt='noop')
+        op1.apply()
+
+        enorm1 = norm(e)
+        gnorm1 = norm(g)
+
+        # reset
+        e.data[:] = 0
+        g.data[:] = 0
+
+        # Coupled
+        # TODO: Need more friendly API for coupled - just
+        # using a dict for now
+        petsc3 = PETScSolve({e: [eq1], g: [eq2]})
+        with switchconfig(openmp=False, mpi=True):
+            op2 = Operator(petsc3, opt='noop')
+        op2.apply()
+
+        enorm2 = norm(e)
+        gnorm2 = norm(g)
+
+        print('enorm1:', enorm1)
+        print('enorm2:', enorm2)
+        assert np.isclose(enorm1, enorm2, rtol=1e-16)
+        assert np.isclose(gnorm1, gnorm2, rtol=1e-16)
+
+        callbacks1 = [meta_call.root for meta_call in op1._func_table.values()]
+        callbacks2 = [meta_call.root for meta_call in op2._func_table.values()]
+
+        # Solving for multiple fields within the same matrix system requires
+        # additional machinery and more callback functions
+        assert len(callbacks1) == 8
+        assert len(callbacks2) == 11
+
+        # Check fielddata type
+        fielddata1 = petsc1[0].rhs.fielddata
+        fielddata2 = petsc2[0].rhs.fielddata
+        fielddata3 = petsc3[0].rhs.fielddata
+
+        assert isinstance(fielddata1, FieldData)
+        assert isinstance(fielddata2, FieldData)
+        assert isinstance(fielddata3, MultipleFieldData)
+
+    @skipif('petsc')
+    def test_coupled_structs(self):
+        grid = Grid(shape=(11, 11))
```

Comment:
This gets used in a lot of tests
```python
grid = Grid(shape=(11, 11))
e = Function(name='e', grid=grid, space_order=2)
f = Function(name='f', grid=grid, space_order=2)
g = Function(name='g', grid=grid, space_order=2)
h = Function(name='h', grid=grid, space_order=2)
```
would a utility function to set these up be worthwhile?

Reply:
Or a list comprehension?